### ISSUE TITLE
Keep the whitespace modifier in `{%-`, `{{-` and `{#-`

### ISSUE THREAD

**Issue by Kocal** [Kocal/prettier-plugin-twig#4]
See https://github.com/ericmorand/twing/pull/321 and https://github.com/ericmorand/twing/pull/326.

Actually there is no way to keep the whitespace modifier in `{%-`, `{{-` and `{#-`.
They respectively _converted_ to `{%`, `{{` and `{#`, and the same behaviour happens for the closing tags.

An other thing, the following code
```twig
hello
{{- world }}
```

is prettified to
```twig
hello{{ world }}
```

because `hello\n` is trimmed at the end by the Twing lexer, see https://github.com/ericmorand/twing/issues/321#issuecomment-476987422.

**Comment by chasegiunta** [comment]
@Kocal is this because, upon output, the whitespace modifier will remove new lines? That’s always been my personal experience. If twing is following suit from actually twig output...

https://github.com/twigphp/Twig/issues/1423

**Comment by Kocal** [comment]
Yes that's the normal behavior for the printer, not for the parser (that builds the AST).

As said in https://github.com/ericmorand/twing/issues/321#issuecomment-476987422, I think the parser should not mutate previous/next text nodes values because of the whitespace modifier.

Trimming whitespaces should be done by the printer IMO.

**Comment by stof** [comment]
Well, twing was build as a port of the Twig engine to be able to render templates. So it is logical that it handles this whitespace-control feature the same way than the original engine does, at the lexer/parser level, rather than at the compiler level (as that's a feature related to how things appear in the source, and this way avoids having to make sure to modify this info properly during any AST transformation to apply the feature in a delayed way).
But that makes the library unsuited for a pretty-printer tool.

**Comment by Kocal** [comment]
> as that's a feature related to how things appear in the source, and this way avoids having to make sure to modify this info properly during any AST transformation to apply the feature in a delayed way

Well, I don't understand why you say it's a feature, for me it looks like a bug.
Why should the parser mutate previous/next text nodes value? That's not the parser's job to do that IMO, let the printer/compiler handle this when compiling the AST to JS code.

Indeed, with this behavior, we can't use Twing _yet_ (I hope :D) to re-build a correct code without trimming and converting `{%-` to `{%`.

I've started to work on a PR (https://github.com/ericmorand/twing/pull/326), actually the lexer is now capable to handle whitespace modifiers, but not the parser, I've tried some thing but it's beyond my capabilities. 😕

**Comment by stof** [comment]
@Kocal The whitespace control rule says "removes the whitespace before/after the tag so they are not part of the output". Moving that responsibility to the compiler would require all other logic (the tag parsers, and all node visitors implementing some optimizations of the AST) to preserve any information needed to know *which* spaces of a previous node should be removed (AST changes might be wrapping or replacing some nodes, which may require moving that info in the tree). And it also makes it harder for code trying to deal with the text nodes, as getting their final content (the one being output) would require checking whether the previous and following nodes have some whitespace trimming.
Plus, comments are not part of the AST either, but their syntax can hold such trimming.
The whitespace control is defined as a lexing-level feature, which is much simpler.

Note that I don't know the Twing internals. I'm one of the maintainers of the original Twig project, not of the Twing JS port.

**Comment by Kocal** [comment]
> Moving that responsibility to the compiler would require all other logic (the tag parsers, and all node visitors implementing some optimizations of the AST) to preserve any information needed to know which spaces of a previous node should be removed 

> The whitespace control is defined as a lexing-level feature, which is much simpler.

Well that's *definitly* true :sweat_smile: 
In Twing's parser and tokens, I've tried to save whitespaces controls in nodes, but that's really a pain in the ass (https://github.com/ericmorand/twing/pull/326/commits/78fa7a6e677709c0c67fdd821bfb8cedeee1ff44) (and it's ugly and does work fully :roll_eyes:).

> And it also makes it harder for code trying to deal with the text nodes, as getting their final content (the one being output) would require checking whether the previous and following nodes have some whitespace trimming.

True. 
Actually in https://github.com/ericmorand/twing/pull/326 I've started to add four new properties to `TwingNodeBlock` (`{% block ...%}{% endblock %}`) to describe whitespaces control behaviors for opening and closing blocks (`{%`) from `block` and `endblock`. But it's really verbose and counterproductive because we should handle two times `{%` and `%}` blocks.

For other _one-element_ nodes, like `TwingNodeDo` (`{% do ... %}`), I would like to add two new properties, but that should be named differently from two-parts blocks.

But that will probably leads to unreadable code and issues... :/

> Plus, comments are not part of the AST either, but their syntax can hold such trimming.

Twing parse comments since https://github.com/ericmorand/twing/pull/329 but does not display them. Maybe the same behavior should be applied on Twig too? Comments are parts of the code, I would be nice to keep them in the AST IMO (and of course not displaying them).

---

Anyway, I don't know what should I do now.
This Prettier plugin can not be continuated and released if it breaks the user's code if he use `{%-`

Maybe I will take severals days to study Twing in deep and have a better knowledge, then trying to fix the problem with a better solution that the actual proposed one.
